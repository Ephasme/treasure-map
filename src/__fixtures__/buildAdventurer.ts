import { IAdventurer } from "../models";
import { North } from "../utils/directions";
import { vector } from "../utils/vector";

export const buildAdventurer = (partialAdventurer?: Partial<IAdventurer>): IAdventurer => ({
    type: "Adventurer",
    location: vector(5, 2),
    traversable: false,
    orientation: North,
    name: "name",
    ...partialAdventurer,
});
