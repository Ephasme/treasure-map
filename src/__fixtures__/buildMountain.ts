import { IMountain } from "../models";
import { vector } from "../utils/vector";

export const buildMountain = (partialMountain?: Partial<IMountain>): IMountain => ({
    type: "Mountain",
    location: vector(5, 2),
    traversable: false,
    ...partialMountain,
});
