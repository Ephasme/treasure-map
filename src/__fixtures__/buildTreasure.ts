import { ITreasure } from "../models";
import { vector } from "../utils/vector";

export const buildTreasure = (partialTreasure?: Partial<ITreasure>): ITreasure => ({
    type: "Treasure",
    location: vector(5, 2),
    traversable: true,
    quantity: 4,
    ...partialTreasure,
});
