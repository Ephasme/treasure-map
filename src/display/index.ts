import { IAdventurer, IMountain, ITreasure } from "../models";
import { GameState } from "../store/state";
import { IVector } from "../utils/vector";
import { IDirection } from "../utils/directions";

function renderDirection(direction: IDirection): string {
    switch (direction.name) {
        case "EAST": return "E";
        case "WEST": return "O";
        case "NORTH": return "N";
        case "SOUTH": return "S";
    }
}

function renderMapSize(vector: IVector): string {
    return `C - ${vector.x} - ${vector.y}\n`;
}

function renderMountain(moutain: IMountain): string {
    const { location: { x, y } } = moutain;
    return `M - ${x} - ${y}\n`;
}

function renderTreasure(treasure: ITreasure): string | undefined {
    const { location: { x, y }, quantity } = treasure;
    if (treasure.quantity > 0) {
        return `T - ${x} - ${y} - ${quantity}\n`;
    }
}

function renderAdventurer(adventurer: IAdventurer): string {
    const {
        name,
        location: { x, y },
        orientation,
        treasures,
    } = adventurer;
    return `A - ${name} - ${x} - ${y} - ${renderDirection(orientation)} - ${treasures}\n`;
}

export function render(state: GameState): string {
    let result = "";
    result += renderMapSize(state.mapSize);
    for (const [, obj] of state.objects) {
        switch (obj.type) {
            case "Adventurer": {
                result += renderAdventurer(obj);
                break;
            }
            case "Mountain": {
                result += renderMountain(obj);
                break;
            }
            case "Treasure": {
                const treasure = renderTreasure(obj);
                if (treasure) {
                    result += treasure;
                }
                break;
            }
        }
    }
    return result;
}
