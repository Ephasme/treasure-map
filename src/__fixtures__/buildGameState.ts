import { Map } from "immutable";
import { AnyObject } from "../models";
import { GameState } from "../store/state";
import { South, West } from "../utils/directions";
import { buildAdventurer } from "./buildAdventurer";
import { buildMountain } from "./buildMountain";
import { buildTreasure } from "./buildTreasure";

export const buildGameState = (gameState?: Partial<GameState>): GameState => ({
    mapSize: {x: 4, y: 4},
    adventurersOrder: [2, 1, 4],
    objects: Map<number, AnyObject>([
        [0, buildMountain({ location: {x: 3, y: 3 } })],
        [1, buildAdventurer({ name: "A", location: {x: 2, y: 3}, orientation: South })],
        [2, buildAdventurer({ name: "B", location: {x: 1, y: 0}, orientation: West })],
        [3, buildTreasure({ location: {x: 0, y: 0}, quantity: 3 })],
        [4, buildAdventurer({ name: "C", location: {x: 3, y: 0}})],
    ]),
    ...gameState,
});
