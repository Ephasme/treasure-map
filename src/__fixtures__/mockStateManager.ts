import { IStore } from "../store/IStore";
import { GameState } from "../store/state";
import { buildGameState } from "./buildGameState";

export function mockStore(gameState?: Partial<GameState>) {
    const dispatch = jest.fn();
    const mGameState = buildGameState(gameState);
    const store: IStore = {
        getState: () => mGameState,
        dispatch,
    };
    return {store, dispatch};
}
