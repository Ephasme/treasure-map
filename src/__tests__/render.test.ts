import { buildGameState } from "../__fixtures__/buildGameState";
import { render } from "../display";

it("should render the final state", () => {
    const result = render(buildGameState());
    expect(result).toMatchSnapshot();
});
