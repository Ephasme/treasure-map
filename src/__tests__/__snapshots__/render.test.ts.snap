// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the final state 1`] = `
"C - 4 - 4
M - 3 - 3
A - A - 2 - 3 - S - 4
A - B - 1 - 0 - O - 4
T - 0 - 0 - 3
A - C - 3 - 0 - N - 4
"
`;
